benchmarks/python/optimized-numpy.py

```diff
@@ -0,0 +1,123 @@
+import sys
+import math
+import timeit
+
+import numpy as np
+import pandas as pd
+
+def load_input_data(path):
+
+    df = pd.read_csv(
+        path, names = ["mass", "x", "y", "z", "vx", "vy", "vz"], delimiter = r"\s+"
+    )
+
+    masses = df["mass"].values.copy()
+    positions = df.loc[:, ["x", "y", "z"]].values.copy()
+    velocities = df.loc[:, ["vx", "vy", "vz"]].values.copy()
+
+    return masses, positions, velocities
+
+def compute_accelerations(accelerations, masses, positions):
+    nb_particles = masses.size
+
+    for index_p0 in range(nb_particles - 1):
+        position0 = positions[index_p0]
+        mass0 = masses[index_p0]
+
+        vectors = position0 - positions[index_p0 + 1: nb_particles]
+
+        distances = np.square(vectors).sum(axis = 1)
+        coefs = distances ** 1.5
```

Comment:
Yes, you are right. Thanks for your suggestion! 
I modified the code in the following ways:
- **The best one:** The time it took: `177.2893s`.
```
coefs = 1. / distances ** 1.5 
accelerations[index_p0] += np.sum((masses[index_p0 + 1: nb_particles] * -1 * vectors.T * coefs).T, axis = 0) 
accelerations[index_p0 + 1: nb_particles] += (mass0 * vectors.T * coefs).T
```
 - The second best one: The time it took: `181.0388s`.
 ```
coefs = np.sqrt(distances ** 3)
_x = np.add(np.sum(
            np.divide(
                np.multiply(
                    masses[index_p0 + 1: nb_particles], -1 * vectors.T
                    ),
                coefs).T, axis = 0),
            accelerations[index_p0]
            )
accelerations[index_p0] = _x

_temp = np.add(
                np.divide(
                np.multiply(
                    mass0, vectors.T
                    ),
                coefs).T,
            accelerations[index_p0 + 1: nb_particles]
            )
accelerations[index_p0 + 1: nb_particles] = _temp
```
- The third one: The time it took: `188.0585s`.
```
coefs = distances ** 1.5
_x = np.add(np.sum(
            np.divide(
                np.multiply(
                    masses[index_p0 + 1: nb_particles], -1 * vectors.T
                    ),
                coefs).T, axis = 0),
            accelerations[index_p0]
            )
accelerations[index_p0] = _x

_temp = np.add(
                np.divide(
                np.multiply(
                    mass0, vectors.T
                    ),
                coefs).T,
            accelerations[index_p0 + 1: nb_particles]
            )
accelerations[index_p0 + 1: nb_particles] = _temp
```
- Not so good one: The time it took: `191.1135s`.
```
coefs = 1. / np.sqrt(distances**3
accelerations[index_p0] += np.sum((masses[index_p0 + 1: nb_particles] * -1 * vectors.T * coefs).T, axis = 0) 
accelerations[index_p0 + 1: nb_particles] += (mass0 * vectors.T * coefs).T
```

These codes are run for `50` iterations for a dataset of size `16`. I'll update it in the PR. Thanks! 
I'm not sure whether it's okay to draw such conclusions for smaller datasets (here, `16`) and a lesser number of iterations (here, `50`).

Reply:
Yes, thanks! I'll improve those codes.

Reply:
I just realized one possible point for confusing fluctuations (*possible* without profiling I have no idea if this matters at all):
```
In [1]: arr = np.random.random((1000, 3))
In [2]: arr2 = arr.copy("F")

In [3]: %timeit arr.sum(-1)
11.1 µs ± 466 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [4]: %timeit arr2.sum(-1)
3.22 µs ± 3.6 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [5]: %timeit arr[:, 0] + arr[:, 1] + arr[:, 2]
3.53 µs ± 49.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
The reason being that summing 3 elements can be slow in NumPy, NumPy is not smart enough to decide that 3 elements is so few that it would make sense to swap the iteration order (in most cases) – nor do we have blocked iteration that might also fix this.

(But I am not sure we should obfuscate the code even if this turns out to make a difference)